File: examples/cpp/navigation_publisher/src/navigation_publisher.cpp
#include "navigation_publisher.hpp"

#include <atomic>
#include <chrono>
#include <cmath>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <thread>

constexpr auto FRAME_RATE = 10;
std::atomic_bool running{true};

void signalHandler(int) {
    std::cerr << "SIGINT or SIGTERM received. Exiting..." << std::endl;
    running = false;
}

int main(int argc, char* argv[]) {
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);

    nodar::zmq::NavigationPublisher publisher;

    std::cout << "Publishing navigation data at " << FRAME_RATE << " Hz" << std::endl;
    std::cout << "Press Ctrl+C to stop..." << std::endl;

    while (running) {
        // Get current timestamp
        const auto timestamp =
            std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now().time_since_epoch())
                .count();

        // Create NavigationData message
        nodar::zmq::NavigationData nav_data;
        nav_data.timestamp_ns = timestamp;

        // IMU data (example values)
        nav_data.imu.timestamp_ns = timestamp;
        nav_data.imu.acceleration_x_m_s2 = 0.0f;
        nav_data.imu.acceleration_y_m_s2 = 0.0f;
        nav_data.imu.acceleration_z_m_s2 = 9.81f;
        nav_data.imu.gyro_x_rad_s = 0.0f;
        nav_data.imu.gyro_y_rad_s = 0.0f;
        nav_data.imu.gyro_z_rad_s = 0.0f;
        nav_data.imu.magnetometer_x_gauss = 0.0f;
        nav_data.imu.magnetometer_y_gauss = 0.0f;
        nav_data.imu.magnetometer_z_gauss = 0.0f;
        nav_data.imu.temperature_degC = 25.0f;

        // GPS data (example values)
        nav_data.gps.timestamp_ns = timestamp;
        nav_data.gps.latitude_deg = 0.0f;
        nav_data.gps.longitude_deg = 0.0f;
        nav_data.gps.altitude_m = 10.0f;
        nav_data.gps.horizontal_uncertainty_m = 0.0f;
        nav_data.gps.vertical_uncertainty_m = 0.0f;
        nav_data.gps.speed_m_s = 5.0f;
        nav_data.gps.course_deg = 0.0f;
        nav_data.gps.fix_type = 0;
        nav_data.gps.num_satellites = 1;

        // Odometry data (example values - vehicle moving forward at 5 m/s in body frame)
        nav_data.odom.timestamp_ns = timestamp;
        nav_data.odom.position_x_m = 0.0f;
        nav_data.odom.position_y_m = 0.0f;
        nav_data.odom.position_z_m = 0.0f;
        nav_data.odom.velocity_x_m_s = 5.0f;  // 5 m/s forward
        nav_data.odom.velocity_y_m_s = 0.0f;  // no lateral motion
        nav_data.odom.velocity_z_m_s = 0.0f;  // no vertical motion
        nav_data.odom.angular_velocity_x_rad_s = 0.0f;
        nav_data.odom.angular_velocity_y_rad_s = 0.0f;
        nav_data.odom.angular_velocity_z_rad_s = 0.0f;

        // Transformation matrix from body frame to Nodar raw camera frame
        // Body frame: x=forward, y=left, z=up
        // Nodar raw camera frame: x=right, y=down, z=forward
        // Transform: Nodar_x = -Body_y, Nodar_y = -Body_z, Nodar_z = Body_x
        nav_data.T_body_to_raw_camera = {
            0.0f, -1.0f, 0.0f,  0.0f,  // Row 1: Nodar x = -Body y
            0.0f, 0.0f,  -1.0f, 0.0f,  // Row 2: Nodar y = -Body z
            1.0f, 0.0f,  0.0f,  0.0f,  // Row 3: Nodar z = Body x
            0.0f, 0.0f,  0.0f,  1.0f  // Row 4: Homogeneous
        };

        if (publisher.publishNavigation(nav_data)) {
            std::cout << "[" << timestamp << "] Publishing NavigationData:" << std::endl;
            std::cout << "  IMU: accel=(" << std::fixed << std::setprecision(2) << nav_data.imu.acceleration_x_m_s2
                      << ", " << nav_data.imu.acceleration_y_m_s2 << ", " << nav_data.imu.acceleration_z_m_s2
                      << ") m/s², gyro=(" << std::setprecision(3) << nav_data.imu.gyro_x_rad_s << ", "
                      << nav_data.imu.gyro_y_rad_s << ", " << nav_data.imu.gyro_z_rad_s << ") rad/s, mag=("
                      << std::setprecision(2) << nav_data.imu.magnetometer_x_gauss << ", "
                      << nav_data.imu.magnetometer_y_gauss << ", " << nav_data.imu.magnetometer_z_gauss
                      << ") gauss, temp=" << std::setprecision(1) << nav_data.imu.temperature_degC << "°C" << std::endl;
            std::cout << "  GPS: lat=" << std::setprecision(6) << nav_data.gps.latitude_deg
                      << "°, lon=" << nav_data.gps.longitude_deg << "°, alt=" << std::setprecision(2)
                      << nav_data.gps.altitude_m << "m, speed=" << nav_data.gps.speed_m_s
                      << "m/s, fix=" << nav_data.gps.fix_type << ", sats=" << nav_data.gps.num_satellites << std::endl;
            std::cout << "  Odom: pos=(" << nav_data.odom.position_x_m << ", " << nav_data.odom.position_y_m << ", "
                      << nav_data.odom.position_z_m << ")m, vel=(" << nav_data.odom.velocity_x_m_s << ", "
                      << nav_data.odom.velocity_y_m_s << ", " << nav_data.odom.velocity_z_m_s << ")m/s, ang_vel=("
                      << std::setprecision(3) << nav_data.odom.angular_velocity_x_rad_s << ", "
                      << nav_data.odom.angular_velocity_y_rad_s << ", " << nav_data.odom.angular_velocity_z_rad_s
                      << ")rad/s" << std::endl;
            std::cout << std::endl;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds((1000 / FRAME_RATE)));
    }

    std::cout << "\nPublisher stopped." << std::endl;
    return EXIT_SUCCESS;
}

File: examples/python/navigation_publisher/navigation_publisher/navigation_publisher.py
import os
import signal
import sys
import time
from datetime import datetime

import zmq

FRAME_RATE = 10

try:
    from zmq_msgs.navigation import NavigationData
    from zmq_msgs.topic_ports import NAVIGATION_TOPIC
except ImportError:
    sys.path.append(
        os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../zmq_msgs/python"))
    )
    from zmq_msgs.navigation_data import NavigationData
    from zmq_msgs.topic_ports import NAVIGATION_TOPIC

# Global variable for signal handling
running = True


def signal_handler(signum, frame):
    global running
    print("\nSIGINT or SIGTERM received. Exiting...", file=sys.stderr)
    running = False


class Publisher:
    def __init__(self, topic, ip, port):
        self.context = zmq.Context(1)
        self.socket = self.context.socket(zmq.PUB)
        self.socket.setsockopt(zmq.SNDHWM, 1)
        if ip:
            self.socket.connect(f"tcp://{ip}:{port}")
        else:
            self.socket.bind(f"tcp://*:{port}")

    def send(self, buffer):
        self.socket.send(buffer)


class NavigationPublisher:
    def __init__(self):
        self.publisher = Publisher(NAVIGATION_TOPIC.name, "", NAVIGATION_TOPIC.port)
        time.sleep(1)

    def publish_navigation(self, nav_data):
        """
        Publish navigation data.

        Args:
            nav_data: NavigationData object
        """
        buffer = bytearray(nav_data.msg_size())
        nav_data.write(buffer, 0)
        self.publisher.send(buffer)
        return True


def main():
    global running

    # Setup signal handlers
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    publisher = NavigationPublisher()

    print(f"Publishing navigation data at {FRAME_RATE} Hz")
    print("Press Ctrl+C to stop...")

    while running:
        # Get current timestamp
        timestamp = int(datetime.utcnow().timestamp() * 1e9)

        # Create NavigationData message
        nav_data = NavigationData()
        nav_data.timestamp_ns = timestamp

        # IMU data (example values)
        nav_data.imu_timestamp_ns = timestamp
        nav_data.imu_acceleration_x_m_s2 = 0.0
        nav_data.imu_acceleration_y_m_s2 = 0.0
        nav_data.imu_acceleration_z_m_s2 = 9.81
        nav_data.imu_gyro_x_rad_s = 0.0
        nav_data.imu_gyro_y_rad_s = 0.0
        nav_data.imu_gyro_z_rad_s = 0.0
        nav_data.imu_magnetometer_x_gauss = 0.0
        nav_data.imu_magnetometer_y_gauss = 0.0
        nav_data.imu_magnetometer_z_gauss = 0.0
        nav_data.imu_temperature_degC = 25.0

        # GPS data (example values)
        nav_data.gps_timestamp_ns = timestamp
        nav_data.gps_latitude_deg = 0.0
        nav_data.gps_longitude_deg = 0.0
        nav_data.gps_altitude_m = 10.0
        nav_data.gps_horizontal_uncertainty_m = 0.0
        nav_data.gps_vertical_uncertainty_m = 0.0
        nav_data.gps_speed_m_s = 5.0
        nav_data.gps_course_deg = 0.0
        nav_data.gps_fix_type = 0
        nav_data.gps_num_satellites = 1

        # Odometry data (example values - vehicle moving forward at 5 m/s in body frame)
        nav_data.odom_timestamp_ns = timestamp
        nav_data.odom_position_x_m = 0.0
        nav_data.odom_position_y_m = 0.0
        nav_data.odom_position_z_m = 0.0
        nav_data.odom_velocity_x_m_s = 5.0  # 5 m/s forward
        nav_data.odom_velocity_y_m_s = 0.0  # no lateral motion
        nav_data.odom_velocity_z_m_s = 0.0  # no vertical motion
        nav_data.odom_angular_velocity_x_rad_s = 0.0
        nav_data.odom_angular_velocity_y_rad_s = 0.0
        nav_data.odom_angular_velocity_z_rad_s = 0.0

        # Transformation matrix from body frame to Nodar raw camera frame
        # Body frame: x=forward, y=left, z=up
        # Nodar raw camera frame: x=right, y=down, z=forward
        # Transform: Nodar_x = -Body_y, Nodar_y = -Body_z, Nodar_z = Body_x
        nav_data.T_body_to_raw_camera = [
            0.0, -1.0, 0.0, 0.0,  # Row 1: Nodar x = -Body y
            0.0, 0.0, -1.0, 0.0,  # Row 2: Nodar y = -Body z
            1.0, 0.0, 0.0, 0.0,  # Row 3: Nodar z = Body x
            0.0, 0.0, 0.0, 1.0  # Row 4: Homogeneous
        ]

        if publisher.publish_navigation(nav_data):
            print(f"[{timestamp}] Publishing NavigationData:")
            print(
                f"  IMU: accel=({nav_data.imu_acceleration_x_m_s2:.2f}, {nav_data.imu_acceleration_y_m_s2:.2f}, {nav_data.imu_acceleration_z_m_s2:.2f}) m/s², "
                f"gyro=({nav_data.imu_gyro_x_rad_s:.3f}, {nav_data.imu_gyro_y_rad_s:.3f}, {nav_data.imu_gyro_z_rad_s:.3f}) rad/s, "
                f"mag=({nav_data.imu_magnetometer_x_gauss:.2f}, {nav_data.imu_magnetometer_y_gauss:.2f}, {nav_data.imu_magnetometer_z_gauss:.2f}) gauss, "
                f"temp={nav_data.imu_temperature_degC:.1f}°C")
            print(f"  GPS: lat={nav_data.gps_latitude_deg:.6f}°, lon={nav_data.gps_longitude_deg:.6f}°, "
                  f"alt={nav_data.gps_altitude_m:.2f}m, speed={nav_data.gps_speed_m_s:.2f}m/s, "
                  f"fix={nav_data.gps_fix_type}, sats={nav_data.gps_num_satellites}")
            print(
                f"  Odom: pos=({nav_data.odom_position_x_m:.2f}, {nav_data.odom_position_y_m:.2f}, {nav_data.odom_position_z_m:.2f})m, "
                f"vel=({nav_data.odom_velocity_x_m_s:.2f}, {nav_data.odom_velocity_y_m_s:.2f}, {nav_data.odom_velocity_z_m_s:.2f})m/s, "
                f"ang_vel=({nav_data.odom_angular_velocity_x_rad_s:.3f}, {nav_data.odom_angular_velocity_y_rad_s:.3f}, {nav_data.odom_angular_velocity_z_rad_s:.3f})rad/s")
            print()

        time.sleep(1.0 / FRAME_RATE)

    print("\nPublisher stopped.")


if __name__ == "__main__":
    main()
